### ISSUE TITLE
Design new API clients using futures

### ISSUE THREAD

**Issue by dermesser** [Byron/google-apis-rs#189]
Background discussed on #188

This probably involves:

- Wait for tokio ecosystem to become stable (reach 1.0?)
- Replacing the python code generator with a Rust one
- Redesigning the templates, of course
- Figure out whether write-everything-in-one-tree is good, or if there's a way to share more code
  - I'm relatively ok with the current approach; the generated crates are not the biggest offenders in terms of compile time or binary size, compared to e.g. syntex_syntax (aaaaggh)
  - Maybe there's a clever way of using build scripts to do these things on the fly, directly from the API definitions.
- Make yup-oauth2 asynchronous, too

Non-goals for now, IMHO:
- support grpc APIs

**Comment by kaendfinger** [comment]
Doing this would be great!

**Comment by thedodd** [comment]
This would be a massive win!

**Comment by dermesser** [comment]
no, Google offering all these APIs as gRPC would be a massive win ;-)

**Comment by dermesser** [comment]
quick update: I looked at the state of futures and tokio, and as of now futures is at 0.2, but tokio still depends on 0.1. And because I don't like working on such a wobbly basis, I consider this still blocked for now...

**Comment by prasannavl** [comment]
Futures are now stable in nightly. 0.3 is chugging along nicely. I wonder what the state of this is, or the library in general.

**Comment by dermesser** [comment]
now with #173 merged it probably makes sense to look at this again. I haven't thought about it in a long time, and am very happy to accept volunteers here as I'm not sure when I find the time to work more on it. But I will at least familiarize myself again, maybe it's not that difficult after all?

**Comment by dermesser** [comment]
In the last few days I put some effort into making yup-oauth2 offer a futures based api... and all I can say is, it's not easy. The entire - relatively transparent - code base is written synchronously, and already the types required to map it into futures are a nightmare. Maybe writing yup-oauth**3** is even easier than porting the current code to Futures :) (in that case, difficult parts like the device flow could be deprioritized in favor or Installed and Service Account flows).

However, and as described above, it is somewhat likely that converting just the google-apis-rs crates to futures is actually easier than moving yup-oauth2, because they are very much request-response based, whereas yup-oauth2 has some asynchronous parts with user feedback etc. Note also that asynchronous yup-oauth2 is not a requirement for google-apis-rs being asynchronous, but it would be very nice - otherwise users would need to mix synchronous and asynchronous code.

I will check how difficult it is to rewrite the easy parts of yup-oauth2 with futures, reusing parts like the crypto utils etc. in order to move forward with google-apis-rs afterwards.

**Comment by dermesser** [comment]
So I've made some good progress on making yup-oauth2 work with futures: https://github.com/dermesser/yup-oauth2/tree/futures

It still lacks a reimplementation of `Authenticator`, but that's relatively doable. All flows (ServiceAccount, Installed, Device) already work with futures/tokio.

Some of the function interfaces are now dumbed down, as ownership is much complicated by futures.

**Comment by dermesser** [comment]
This is mostly finished (for now), although I expect some minor improvements to types/function signatures that solve some problems I've had so far.

Now the exciting part is coming up, writing the API code with futures :) Due to all the retry logic it will not be trivial to convert, but on the upside, the only code that actually needs to be ported seems to be contained in the `doit()` methods of the various call builders. The basic structure is already similar to what happens in `yup-oauth2`, so it might end up being doable while mostly preserving the current API (as opposed to my adventurous idea above).

e.g.

```rust
pub fn doit(mut self) -> Result<(hyper::client::Response, Permission)> { ... }
```

becomes

```rust
pub fn doit(mut self) -> impl Future<Item=(hyper::client::Response,Permission), Error=cmn::Error> { ... }
```

@Byron are you having any free capacity in the next weeks/months? The mako templates are intimidating me quite a bit so far :)

**Comment by dermesser** [comment]
note: in dermesser/yup-oauth2#94, I merged the futures based OAuth implementation. It is now available on crates.io as v3.0.

**Comment by Byron** [comment]
@dermesser Today I am taking all the time it takes to catch up on maintaining the various open source projects of mine, and I will at least plow through the issues, make minor fixes and re-release all crates.

Whenever I look at the Mako templates, I think 'OMG, let me rewrite this with purely declarative templates driven by data produced by Rust code', but then again, it's relatively straightforward to use once you have a syntax highlighter set. Then it's mainly the `change mako -> make apiname-cli-carg ARGS=check` loop to get anywhere.

Remembering the past, I usually sorted APIs by size and worked on the smallest one, and once that worked, moved to bigger APIs until all the kinks were etched out.

What I am most afraid of is the futures themselves, knowing that compile errors result in illegible types. AFAIK the compiler still doesn't highlight only the mismatching portion of the type.

Thinking back and forth, I believe in the end it's easier not to rewrite the whole templating portion just to migrate to async. The former would certainly be intriguing, but essentially is a rewrite that can fix all issues this repository has, and be the official successor when done.

Until then, I guess we will have to live with Python, and maybe value it for its 'hackability' :D.

Also CC @TheBiggerGuy as he has quite some experience with the templating system.

**Comment by application-developer-DA** [comment]
> What I am most afraid of is the futures themselves, knowing that compile errors result in illegible types. AFAIK the compiler still doesn't highlight only the mismatching portion of the type.

Even if a trick with `impl Future<...>` does not work (like for instance if you have 2 branches of code, each of them has a different type, or in case if you use trait which does not allow using `impl Trait` syntax inside their method declarations), you can always use the `Box<dyn Future<Item=X, Error=Y> + Send + 'static>` syntax.

I think it's great that there are bindings for Google APIs as it seems that the rest of the crates provide access via obsolete methods and still lack the "full compatibility". These `google-*` crates seem to be more up to date, but unfortunately they're not very useful on a backend without the support of futures as you can't combine them with tokio that well.

**Comment by dermesser** [comment]
@application-developer-DA fundamentally I agree with you, although @Byron's concerns are valid. Without wanting to scare anyone, I've spent a few hours not even trying to write "nice" futures code, but just trying to understand what's wrong while migrating the yup-oauth2 code :) (and I was very happy once I got partially rid of `Box<dyn Future<Item=X, Error=Box<dyn Error + Send> + Send>`)
At first glance it seems like the code involved here may be simpler because there's not that much interaction within the library, and you can use a specific future for every doit() method (but I can not be sure, because I don't know it well enough)

**Comment by fzzzy** [comment]
Hello,

I have already partially updated google-apis-rs to use Future.

https://github.com/fzzzy/google-apis-rs/tree/wip-update-to-hyper-0.12

However, since the version of yup-oauth2 I was depending on called wait on a Future to continue present a synchronous interface, but we are running inside the executor already, it deadlocks in the call to wait. I attempted to convert yup-oauth2 to return Futures but found it to be difficult and I took a different tack for our project temporarily (using a threadpool to make google-apis-rs async instead).

Now that yup-oauth2 has been converted to return Futures (thanks!), I believe it should be relatively simple to finish the wip-update-to-hyper-0.12 branch. However, I don't have time at the moment. Is someone willing to take it over from me?

**Comment by fzzzy** [comment]
@dermesser I believe you are correct; the only change to the api required is that doit will now return a Future, and every doit has a unique Future type.

**Comment by Byron** [comment]
A rewrite of google-apis [is on the way](https://github.com/google-apis-rs/generator/projects/1), making good progress.
Suffice to say that async is planned to be added once stabilized, and that we still have to sort out the way HTTP clients are supported. For now, it is hard-coded to `request`.

**Comment by prasannavl** [comment]
@jroesch @Byron - You might want to look into https://github.com/rustasync/surf - Rust WG's fully async client.

**Comment by Byron** [comment]
This is done with the latest published version of the crates: v2.0.
